File: 프로그래머스/1/42889. 실패율/실패율.py
def solution(N, stages):
    stage_count = [0] * (N+2)
    
    total_player = len(stages)  

    for stage in stages: 
        stage_count[stage] += 1

    fail_rates = []  
    for i in range(1, N+1): 
        if total_player == 0: 
            fail_rates.append((i, 0))
        else:
            fail_rate = stage_count[i] / total_player  
            fail_rates.append((i, fail_rate)) 
            total_player -= stage_count[i]  
    answer = [stage for stage, rate in sorted(fail_rates, key=lambda x: (-x[1], x[0]))]
    
    return answer

File: 프로그래머스/1/42889. 실패율/실패율.java
import java.util.*;

class Solution {
    public int[] solution(int N, int[] stages) {
        int[] answer = new int[N];
        int[] num = new int[N+2];
        HashMap<Integer, Float> map = new HashMap<>();
        
        for(int stage : stages){
            num[stage] ++;
        }
        
        int length = stages.length;
        for(int i=1; i<=N; i++){
            if (length == 0){
                map.put(i, 0f);
                continue;
            }
            float fail = (float) num[i] / length;
            map.put(i, fail);
            length -= num[i];
        }
        
        ArrayList<Integer> result = new ArrayList<>(map.keySet());
        result.sort(new Comparator<Integer>() {
            public int compare(Integer o1, Integer o2) {
                return Float.compare(map.get(o2), map.get(o1));
            }
        });

        
        for(int i=0; i<N; i++){
            answer[i] = result.get(i);
        }
        
        return answer;
    }
}